lib/elixir/lib/module/parallel_checker.ex

```diff
@@ -0,0 +1,268 @@
+  defp start_and_wait(server) do
+    :gen_server.call(server, :start_and_wait, :infinity)
+  end
+
+  defp lock(server, module) do
+    :gen_server.call(server, {:lock, module}, :infinity)
+  end
+
+  defp unlock(server, module) do
+    :gen_server.call(server, {:unlock, module})
+  end
+
+  defp preload_cache(ets, modules) do
+    Enum.each(modules, fn {map, _binary} ->
+      exports = [{{:__info__, 1}, :def} | definitions_to_exports(map.definitions)]
+      deprecated = :maps.from_list(map.deprecated)
+      cache_module(ets, map.module, exports, deprecated)
+    end)
+  end
+
+  defp spawn_checkers(%{modules: []} = state) do
+    state
+  end
+
+  defp spawn_checkers(%{spawned: spawned, schedulers: schedulers} = state)
+       when spawned >= schedulers do
+    state
+  end
+
+  defp spawn_checkers(%{modules: [{map, binary} | modules]} = state) do
+    parent = self()
+    ets = state.ets
+
+    spawn_link(fn ->
+      {binary, warnings} = Module.Checker.verify(map, binary, {parent, ets})
+      send(parent, {__MODULE__, map.module, binary, warnings})
+    end)
+
+    spawn_checkers(%{state | modules: modules, spawned: state.spawned + 1})
+  end
+
+  defp cache_module({server, ets}, module) do
+    if lock(server, module) do
+      cache_from_chunk(ets, module) || cache_from_info(ets, module)
+      unlock(server, module)
+    end
+  end
+
+  defp cache_from_chunk(ets, module) do
+    with {^module, binary, _filename} <- :code.get_object_code(module),
+         {:ok, chunk} <- get_chunk(binary),
+         {:elixir_checker_v1, map} <- :erlang.binary_to_term(chunk) do
+      exports = :maps.put({:__info__, 1}, :def, map.exports)
+      cache_module(ets, module, exports, map.deprecated)
+      true
+    else
+      _ -> false
+    end
+  end
+
+  defp get_chunk(binary) do
+    case :beam_lib.chunks(binary, ['ExCk'], [:allow_missing_chunks]) do
+      {:ok, {_module, [{'ExCk', :missing_chunk}]}} -> :error
+      {:ok, {_module, [{'ExCk', chunk}]}} -> {:ok, chunk}
+      :error -> :error
+    end
+  end
+
+  defp cache_from_info(ets, module) do
+    if Code.ensure_loaded?(module) do
+      exports = info_exports(module)
+      deprecated = info_deprecated(module)
+      cache_module(ets, module, exports, deprecated)
+    else
+      :ets.insert(ets, {{:cached, module}, false})
+    end
+  end
+
+  defp info_exports(module) do
+    :maps.from_list(
+      Enum.map(module.__info__(:macros), &{&1, :defmacro}) ++
+        Enum.map(module.__info__(:functions), &{&1, :def})
+    )
+  rescue
+    _ -> :maps.from_list(Enum.map(module.module_info(:exports), &{&1, :def}))
+  end
+
+  defp info_deprecated(module) do
+    :maps.from_list(module.__info__(:deprecated))
+  rescue
+    _ -> %{}
+  end
+
+  defp cache_module(ets, module, exports, deprecated) do
+    all_exports =
+      Enum.map(exports, fn {{fun, arity}, kind} ->
+        reason = :maps.get({fun, arity}, deprecated, nil)
+        :ets.insert(ets, {{:export, {module, fun, arity}}, kind, reason})
```

Comment:
If we change the format of the chunk, as I mentioned earlier, then this becomes simpler too, because the format in the chunk will be closer to what we actually store in the ETS table.